### PR TITLE
[Inductor Cutlass backend] Improve GEMM op filtering

### PR DESCRIPTION
**Pull request by kadeng**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #121492
* #124930
* #124929
* #124928
* #124577
* __->__ #124576

Add configurable allowlist / denylist regular expressions to make it possible to exclude certain
CUTLASS GEMM implementations ( for example "pingpong" Kernels due to undesired numerical behavior ).

Remove usage of old 2.x Cutlass Kernels entirely.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by kadeng** [comment]
The profiler test failure appears unrelated..

File: torch/_inductor/config.py [hunk 1]
```diff
@@ -787,6 +787,13 @@ class cuda:
     # epilogue fusion via CUTLASS Epilogue Visitor Trees ( EVT )
     # are enabled for the CUTLASS backend.
     cutlass_only_evt_capable_ops: bool = False
+    # Keep only Cutlass op configs which contain this regular expression pattern
+    # Set this to "warpspecialized_cooperative_epi_tma" to enable only SM90 TMA Cutlass Kernels for large GEMMs
+    cutlass_op_whitelist_regex: Optional[str] = None
+
+    # Filter Cutlass configs which contain this regular expression pattern
+    # Set this to "pingpong" to avoid nondeterministic numerical issues
+    cutlass_op_blacklist_regex: Optional[str] = "pingpong"
```
Comment (jansel) [inline torch/_inductor/config.py:796]: Change whitelist/blacklist to allowlist/denylist.

https://www.splunk.com/en_us/blog/learn/blacklist-whitelist-inclusivity.html

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/config.py (hunk 1 above)
Comment (kadeng) [reply torch/_inductor/config.py]: done

File: test/inductor/test_cutlass_backend.py [hunk 2]
```diff
@@ -425,6 +423,8 @@ def compare_results(
                 "max_autotune_gemm_backends": max_autotune_gemm_backends,
                 "cuda.cutlass_dir": _CUTLASS_DIR,
                 "cuda.cutlass_max_profiling_configs": 4,
+                "cuda.cutlass_op_allowlist_regex": "",
+                "cuda.cutlass_op_denylist_regex": "pingpong",  # Pingpong Kernels can lead to numerical issues
```
Comment (eellison) [inline test/inductor/test_cutlass_backend.py:427]: is there documentation anywhere of the potential ops ? where is pingpong coming from ?

File: torch/_inductor/codegen/cuda/gemm_template.py [hunk 3]
```diff
@@ -475,7 +487,16 @@ def filter_op(
         # Set epilogue.
         # TODO: update epilogue functor according to epilogues.
         op.element_epilogue = op.accumulator_type()
-
+        if inductor_cuda_config.cutlass_op_allowlist_regex is not None:
+            if not re.search(
+                inductor_cuda_config.cutlass_op_allowlist_regex, op.configuration_name()
+            ):
```
Comment (eellison) [inline torch/_inductor/codegen/cuda/gemm_template.py:493]: nit: should we compile the regex once and cache?

File: torch/_inductor/config.py [hunk 4]
```diff
@@ -787,6 +787,13 @@ class cuda:
     # epilogue fusion via CUTLASS Epilogue Visitor Trees ( EVT )
     # are enabled for the CUTLASS backend.
     cutlass_only_evt_capable_ops: bool = False
+    # Keep only Cutlass op configs which contain this regular expression pattern
+    # Set this to "warpspecialized_cooperative_epi_tma" to enable only SM90 TMA Cutlass Kernels for large GEMMs
+    cutlass_op_allowlist_regex: Optional[str] = None
+
+    # Filter Cutlass configs which contain this regular expression pattern
+    # Set this to "pingpong" to avoid nondeterministic numerical issues
+    cutlass_op_denylist_regex: Optional[str] = "pingpong"
```
Comment (eellison) [inline torch/_inductor/config.py:796]: more details or links to docs here might be useful

File: torch/_inductor/kernel/mm.py [hunk 5]
```diff
@@ -284,13 +290,17 @@ def tuned_addmm(inp, mat1, mat2, *, alpha=1, beta=1, layout=None):
...
-            choices,
-            layout,
-            [mat1, mat2, inp_expanded],
-            alpha=alpha,
-            beta=beta,
-        )
+        if (
+            WrapperCodeGen.statically_known_int_or_none(inp_expanded.layout.stride[-1])
+            != 0
+        ):
```
Comment (eellison) [inline torch/_inductor/kernel/mm.py:296]: is this only an issue for this particular cutlass gemm? what about the others?

**Review by eellison — APPROVED** [review]
add tests ?

File: test/inductor/test_cutlass_backend.py (hunk 2 above)
Comment (kadeng) [reply test/inductor/test_cutlass_backend.py:427]: Well, that's a Cutlass internal thing how Kernels are named. New Kernels are being introduced with new Versions of Cutlass all the time. A current list could be generated by calling op.configuration_name() on all Cutlass ops, for example those returned from cutlass_utils.gen_ops(): https://github.com/pytorch/pytorch/blob/438388245c1c98607d8667a617c6af17e5438fef/torch/_inductor/codegen/cuda/cutlass_utils.py#L173  - The Kernels with "pingpong" in their name seems to use a certain zig-zag memory access pattern (thereby the name pingpong ), which in my experience often led to stronger numerical discrepancies with ATen / eager ops for larger GEMMs where many numbers are being added.

File: torch/_inductor/kernel/mm.py (hunk 5 above)
Comment (kadeng) [reply torch/_inductor/kernel/mm.py:296]: I don't know why, but I have seen CUDA illegal memory access errors on recent versions of Cutlass if this dimension had a stride of zero. It's a rare case and most likely a bug in some Cutlass Kernel(s). It's only relevant for the bias in addmm, so yes, the others don't have this issue.

Comment (kadeng) [reply torch/_inductor/kernel/mm.py:296]: Added an explanatory comment.

**Comment by kadeng** [comment]
> add tests ?

Done

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**